### PR TITLE
GH-43860: [Go][Parquet] Handle the error correctly

### LINKED ISSUE #43860: [Go] Error was not handled
### Describe the bug, including details regarding any error messages, version, and platform.

The error thrown during the [ReadRecords](https://github.com/apache/arrow/blob/6b268f62a8a172249ef35f093009c740c32e1f36/go/parquet/file/record_reader.go#L569) process was not handled or returned correctly.
1. `HasNext` returns false because of an error, then the `for` loop broke:

2. But the error isn't being returned:

### Component(s)

Go

### PR DESCRIPTION
**Pull request by bigsheeper**
### Rationale for this change
Fixes: https://github.com/apache/arrow/issues/43860

### What changes are included in this PR?
Return error correctly

### Are these changes tested?
Yes

### Are there any user-facing changes?
Nope

* GitHub Issue: #43860

### CODE REVIEW

**Comment by mapleFU** [comment]
> Are these changes tested?
Yes

Would you mind add a test for this? I guess parquet-testing/bad-data might contain some bad file?

**Comment by bigsheeper** [comment]
> > Are these changes tested?
> > Yes
> 
> Would you mind add a test for this? I guess parquet-testing/bad-data might contain some bad file?

@mapleFU Sorry, could you clarify what you mean by "parquet-testing/bad-file"? Can you provide a link or more context?

**Comment by mapleFU** [comment]
You can try to follow[1]. And parquet-testing uses this repo [2]. It includes some bad data-files in "bad_data", would you mind check can this being tested using these files?

[1] https://github.com/apache/arrow/blob/6b268f62a8a172249ef35f093009c740c32e1f36/go/parquet/file/file_reader_test.go#L397-L401
[2] https://github.com/apache/parquet-testing

**Comment by bigsheeper** [comment]
> You can try to follow[1]. And parquet-testing uses this repo [2]. It includes some bad data-files in "bad_data", would you mind check can this being tested using these files?
> 
> [1]
> 
> https://github.com/apache/arrow/blob/6b268f62a8a172249ef35f093009c740c32e1f36/go/parquet/file/file_reader_test.go#L397-L401
> 
> 
> [2] https://github.com/apache/parquet-testing


@mapleFU It seems that the `PARQUET_TEST_DATA` environment variable only leads to the files from the [data directory](https://github.com/apache/parquet-testing/tree/master/data) and does not include bad files.

**Comment by mapleFU** [comment]
Sorry, would you mind try something like below?

```c++
std::string get_bad_data_dir() {
  // PARQUET_TEST_DATA should point to ARROW_HOME/cpp/submodules/parquet-testing/data
  // so need to reach one folder up to access the "bad_data" folder.
  std::string data_dir(get_data_dir());
  std::stringstream ss;
  ss << data_dir << "/../bad_data";
  return ss.str();
}
```

If cannot touch data, can we handle write a test ourselve using go?

**Comment by mapleFU** [comment]
Both is ok for me. I'll be glad to see a go ut

**Comment by bigsheeper** [comment]
> Both is ok for me. I'll be glad to see a go ut, if bad_data can reproduce it would be convient. I'm ok for both

@mapleFU I attempted to read a bad file, but no error was returned. In my case, the io.reader reported an error (specifically, a timeout connecting to S3: Get "http://x.x.x.x:9000/a-bucket//tmp/test_2292796098596731149.parquet": dial tcp x.x.x.x:9000: connect: connection refused), but the Parquet reader did not propagate this error. I'v simulate this scenario in the ut. Please help to review. :)

File: go/parquet/file/file_reader_test.go [hunk 1]
```diff
@@ -452,6 +452,55 @@ func TestRleBooleanEncodingFileRead(t *testing.T) {
...
+	assert.NoError(t, err)
+
+	fileReader, err := pqarrow.NewFileReader(r, pqarrow.ArrowReadProperties{}, memory.DefaultAllocator)
+	assert.NoError(t, err)
+
+	columnReader, err := fileReader.GetColumn(context.Background(), 0)
+	assert.NoError(t, err)
+
+	_, err = columnReader.NextBatch(1)
+	assert.Error(t, err) // Expect an error to occur.
```
Comment (joellubi) [inline go/parquet/file/file_reader_test.go:501]:
```suggestion
	assert.ErrorContains(t, err, "mock error") // Expect an error to occur.
```

Minor suggestion to make it clear that the intended error was the cause here.

**Comment by mapleFU** [comment]
Dial a s3 address in parquet ut would be a disaster and be unstable, I don't suggest that

Mocking might be ok here

**Comment by bigsheeper** [comment]
> Dial a s3 address in parquet ut would be a disaster and be unstable, I don't suggest that
> 
> Mocking might be ok here

I didn’t dial an S3 address in Parquet ut; I only mock an error when `io.Reader` calls `ReadAt`. 
The connection timeout to S3 is a real issue I encountered in system. We need to ensure this error is thrown and handled properly, otherwise users might face situations where they read 0 rows of Parquet data, which would be disastrous for them.

File: go/parquet/file/file_reader_test.go (hunk 1 above)
Comment (bigsheeper) [reply go/parquet/file/file_reader_test.go]: updated

**Comment by mapleFU** [comment]
> I didn’t dial an S3 address in Parquet ut; I only mock an error when io.Reader calls ReadAt.
We need to ensure any error is thrown and handled properly, otherwise users might face situations where they read 0 rows of Parquet data, which would be disastrous for them.

I see. A lot of module under Parquet-go might need handle `error` gracefully. Maybe some fuzzing script to generate bad file or inject bad io-state would help finding them, sigh...

**Comment by bigsheeper** [comment]
If you're interested in the issue we've encountered, see: https://github.com/milvus-io/milvus/issues/35662 :)

**Comment by zeroshade** [comment]
@mapleFU Several areas of the Go parquet lib rely on using panics with a top-level recover to return errors. I agree we should definitely try to find any spots where we're missing proper handling.

Go has support for fuzz testing built in https://go.dev/doc/security/fuzz/ and I've been meaning to get around to adding that to the parquet lib but haven't had the time. It would be great for us to add that

**Review by zeroshade — APPROVED** [review]
LGTM thanks!

If there are any other areas we can identify that have incorrect error handling we should try to find them, test, and update them.

**Comment by mapleFU** [comment]
@bigsheeper FYI https://github.com/apache/arrow/pull/43607 these are similiar problem

**Review by joellubi — APPROVED** [review]
Thanks for the fix!

**Pull request merged**